src/kafka-net/ProtocolGateway.cs

```diff
@@ -0,0 +1,139 @@
+                    //find route it can chage after Metadata Refresh
+                    var route = _brokerRouter.SelectBrokerRouteFromLocalCache(topic, partition);
+                    var responses = await route.Connection.SendAsync(request);
+                    response = responses.FirstOrDefault();
+                    if (response == null)
+                    {
+                        return null; /*this can happened if you send ProduceRequest with ack level=0 */
+                    }
+
+                    var error = (ErrorResponseCode) response.Error;
+                    if (error == ErrorResponseCode.NoError)
+                    {
+                        return response;
+                    }
+
+                    //It means we had an error 
+                    needToRefreshTopicMetadata = CanRecoverByRefreshMetadata(error);
+
+                }
+                catch (ResponseTimeoutException ex)
+                {
+                    exception = ExceptionDispatchInfo.Capture(ex);
+                    needToRefreshTopicMetadata = true;
+                }
+                catch (SocketException ex)
+                {
+                    exception = ExceptionDispatchInfo.Capture(ex);
+                    needToRefreshTopicMetadata = true;
+                }
+                bool hasMoreRetry = retryTime + 1 < _maxRetry;
+                if (needToRefreshTopicMetadata && hasMoreRetry)
+                {
+                    //Log Retry Sesion id
+                    retryTime++;
+                    await TryRefreshTopicMetadata(topic, topicMetadataRefreshDate, exception, response);
+                }
+                else
+                {
+                    throwError(exception, response);
+                }
+
+            }
+            throw new KafkaApplicationException("FetchResponse returned error condition.  ErrorCode:{0}", response.Error);
+        }
+
+        private static bool CanRecoverByRefreshMetadata(ErrorResponseCode error)
+        {
+
+            return error == ErrorResponseCode.BrokerNotAvailable ||
+                                         error == ErrorResponseCode.ConsumerCoordinatorNotAvailableCode ||
+                                         error == ErrorResponseCode.LeaderNotAvailable ||
+                                         error == ErrorResponseCode.NotLeaderForPartition;
+
+        }
+
+        private async Task TryRefreshTopicMetadata(string topic, DateTime lastTopicMetadataRefreshDate,
+            ExceptionDispatchInfo socketException, IBaseResponse response)
+        {
+            bool metadataNotExpire = !await _brokerRouter.RefreshTopicMetadata(topic);//Do Not Change this order
+            bool metadataNotUpdate = _brokerRouter.GetTopicMetadataRefreshTime(topic) == lastTopicMetadataRefreshDate;
+    
+
+
+            if (metadataNotUpdate && metadataNotExpire)
+            {
+                throwError(socketException, response);
```

Comment:
made change now starting with  capital letter.
if two thing need the same logic it should be a method.